server/services/context.py

```diff
@@ -0,0 +1,104 @@
+"""This module provides a service to determine position
+context inside an XML document."""
+
+from typing import Optional
+from .xsd.types import XsdTree, XsdNode
+
+
+class XmlContext:
+    """Represents the context at a given XML document position.
+
+    It provides information about the current node and element
+    under the cursor.
+    """
+
+    element_name: str
+    node: XsdNode
+    is_empty: bool
+
+    def __init__(self, name: str = None, node: XsdNode = None):
+        self.element_name = name
+        self.node = node
+        self.is_empty = False
+
+
+class XmlContextService:
+    """This service provides information about the XML context at
+    a specific position of the document.
+    """
+
+    xsd_tree: XsdTree
+
+    def __init__(self, xsd_tree: XsdTree):
+        self.xsd_tree = xsd_tree
+
+    def get_xml_context(self, xml_content: str, offset: int) -> XmlContext:
+        """Gets the XML context at a given offset inside the document.
+
+        Args:
+            xml_content (str): The XML contents.
+            offset (int): The character offset inside de document.
+
+        Returns:
+            XmlContext: The resulting context with the current node
+            definition and other information. If the context can not be
+            determined, the default context with no information is returned.
+        """
+        context = XmlContext()
+        if self.is_empty_content(xml_content):
+            context.node = self.xsd_tree.root
+            context.is_empty = True
+            return context
+
+        current_tag = self.find_current_tag(xml_content, offset)
+        if current_tag:
+            node = self.xsd_tree.find_node_by_name(current_tag)
+            return XmlContext(current_tag, node)
+        return context
+
+    @staticmethod
```

Comment:
Just utility functions that are related but independent of the **state** of the class so I thought it would be better to have them as static methods (at least that is what I would do in C# 😛)